worker/proxyupdater/proxyupdater.go

```diff
@@ -79,36 +77,27 @@ var NewWorker = func(config Config) (worker.Worker, error) {
 	return w, nil
 }
 
-func (w *proxyWorker) writeEnvironmentFile() error {
-	// Writing the environment file is handled by executing the script:
-	//
-	// On cloud-instance ubuntu images, the ubuntu user is uid 1000, but in
-	// the situation where the ubuntu user has been created as a part of the
-	// manual provisioning process, the user will exist, and will not have the
-	// same uid/gid as the default cloud image.
-	//
-	// It is easier to shell out to check, and is also the same way that the file
-	// is written in the cloud-init process, so consistency FTW.
-	filePath := path.Join(w.config.Directory, w.config.Filename)
-	result, err := exec.RunCommands(exec.RunParams{
-		Commands: fmt.Sprintf(
-			`[ -e %s ] && (printf '%%s\n' %s > %s && chown ubuntu:ubuntu %s)`,
-			w.config.Directory,
-			utils.ShQuote(w.proxy.AsScriptEnvironment()),
-			filePath, filePath),
-		WorkingDir: w.config.Directory,
-	})
-
-	if err != nil {
-		return err
+func (w *proxyWorker) saveProxySettingsToFiles() error {
+	// The proxy settings are (usually) stored in three files:
+	// - /etc/juju-proxy.conf - in 'env' format
+	// - /etc/systemd/system.conf.d/juju-proxy.conf
+	// - /etc/systemd/user.conf.d/juju-proxy.conf - both in 'systemd' format
+	for _, file := range w.config.EnvFiles {
+		err := ioutil.WriteFile(file, []byte(w.proxy.AsScriptEnvironment()), 0644)
+		if err != nil {
+			logger.Errorf("Error updating environment file %s - %v", file, err)
+		}
 	}
-	if result.Code != 0 {
-		logger.Errorf("failed writing new proxy values: \n%s\n%s", result.Stdout, result.Stderr)
+	for _, file := range w.config.SystemdFiles {
+		err := ioutil.WriteFile(file, []byte(w.proxy.AsSystemdDefaultEnv()), 0644)
+		if err != nil {
+			logger.Errorf("Error updating systemd file %s - %v", file, err)
```

Comment:
The error from WriteFile will contain the file name, so probably no need to duplicate it: